cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/ComparisonWithWiderType/test3.cpp

```diff
@@ -0,0 +1,3 @@
+void test_issue_5850(unsigned char small, unsigned int large1) {
+	for(; small < static_cast<unsigned char>(large1 - 1); small++) { } // GOOD
```

Comment:
I've convinced myself I'm happy with this.  There _could_ still be a correctness issue if `large1 - 1` overflows an `unsigned char` causing the loop to end at an unexpected point, but (1) the programmer has explicitly made the cast, suggesting they are happy it fits and (2) regardless the loop will terminate (which it might not without the cast), and I remember that being the main concern here.

Reply:
Good point. Would you be happier with the change if we only did `large.getExplicitlyConverted()` instead of `large.getExplicitlyConverted()`? That would only exclude the ones the conversions programmer actually wrote?

Reply:
I think I'm happy with `getFullyConverted` ... LGTM diff to be sure: https://lgtm.com/query/3711128432146950375/

Reply:
Yeah, I'm trying to see if I can produce a testcase that has the problem. Let's hold off with the merge until we have a conclusion on that.

Reply:
It makes sense that fully and explicitly converted are the same on `large`. The query is about a small type compared to a large type, so the implicit conversions will (always?) happen on the small side of the comparison.

Reply:
Oh, yes. That's a good point!

Reply:
Unlikely to be found like this in the wild, but maybe:
```
void test(unsigned char small, unsigned int large1) {
  for(; small < (unsigned int)(unsigned char)(large1 - 1); small++) { } // GOOD
}
```